### PR TITLE
Don't serialize all variant override objects where they are not needed

### PR DESCRIPTION
**Pull request by Matt-Yorkley**
#### What? Why?

This bit of data is only needed in one place, and only as a count (not all the objects), and can cause huge amounts of superfluous data to be fetched and serialized, for example in the admin products index page (where variant override data is not used at all). So for example, a variant could have 20 different overrides (from different distributors), and every override would be serialized in full, for every variant, for every product. And they're not even used... :boom:

Noticed whilst investigating #6608

#### What should we test?

Affects admin bulk products page and admin product edit pages. Shouldn't have any noticeable effect on behavior. Should improve performance where lots of products have lots of overrides.

#### Release notes

Changelog Category: Technical changes

### CODE REVIEW

File: app/views/spree/admin/products/index/_products_variant.html.haml [hunk 1]
```diff
@@ -29,6 +29,6 @@
   %td.actions
     %a{ 'ng-click' => 'editWarn(product,variant)', :class => "edit-variant icon-edit no-text", 'ng-show' => "variantSaved(variant)", 'ofn-with-tip' => t(:edit)  }
   %td.actions
-    %span.icon-warning-sign{ 'ng-if' => 'variant.variant_overrides', 'ofn-with-tip' => "{{ 'spree.admin.products.index.products_variant.variant_has_n_overrides' | t:{n: variant.variant_overrides.length} }}" }
+    %span.icon-warning-sign{ 'ng-if' => 'variant.variant_overrides_count > 0', 'ofn-with-tip' => "{{ 'spree.admin.products.index.products_variant.variant_has_n_overrides' | t:{n: variant.variant_overrides_count} }}" }
```
Comment (sauloperez) [inline app/views/spree/admin/products/index/_products_variant.html.haml:32]: it'd be even better with https://apidock.com/rails/ActiveRecord/Base/exists%3f/class

Comment (Matt-Yorkley) [reply app/views/spree/admin/products/index/_products_variant.html.haml:32]: I don't understand how that would be used..?

At this point in the view, the value is an integer that was serilialized previously and is now in a javascript object, we're not querying the actual variant record.

Comment (sauloperez) [reply app/views/spree/admin/products/index/_products_variant.html.haml:32]: nevermind, I missed this was a JS object and not an AR model.

File: app/serializers/api/admin/variant_serializer.rb [hunk 2]
```diff
@@ -5,11 +5,9 @@ module Admin
     class VariantSerializer < ActiveModel::Serializer
       attributes :id, :name, :producer_name, :image, :sku, :import_date,
                  :options_text, :unit_value, :unit_description, :unit_to_display,
-                 :display_as, :display_name, :name_to_display,
+                 :display_as, :display_name, :name_to_display, :variant_overrides_count,
                  :price, :on_demand, :on_hand, :in_stock, :stock_location_id, :stock_location_name
 
-      has_many :variant_overrides
```
Comment (sauloperez) [inline app/serializers/api/admin/variant_serializer.rb:11]: tests are passing without this so we might be lucky :pray:

Comment (Matt-Yorkley) [reply app/serializers/api/admin/variant_serializer.rb:11]: It's only used in the bulk products pages, and they don't use overrides :+1:

**Review by sauloperez — CHANGES_REQUESTED** [review]
I think this warning icon needs a feature spec. I've been digging around and couldn't find any. Perhaps you do @Matt-Yorkley ? We should add one if not. Then, if @filipefurtad0 approves it later on, we don't need to test this manually anymore.

**Comment by Matt-Yorkley** [comment]
> I think this warning icon needs a feature spec.

Good idea :+1: I think it might be on the admin bulk products page and shows a little symbol, I'll take a look.

**Comment by sauloperez** [comment]
yes, it's that warning icon :point_down: 

Screenshot from 2021-01-13 12-47-41

**Review by sauloperez — APPROVED** [review]
:heart:

**Review by mkllnk — APPROVED** [review]
Nice catch. The testing instructions could be a bit more specific but I'm sure that Filipe will think of a good way of testing it.

**Comment by filipefurtad0** [comment]
Hey @Matt-Yorkley ,

 I do see a minor change in the UI:
- after this PR, the warning sign only appears, if there is a variant override (i.e., n=1)
- if n=0, than the warning symbol is not there - which makes sense :+1: 

I found this snippet in the code, so I guess this intended, right? 
`'ng-if' => 'variant.variant_overrides_count > 0'`

The feature spec covers only the case of n=1, but I think it's fine. One could consider adding the case for zero overrides, in which it would be expected (after this PR) that the warning symbol does not appear at all (for n=0 variants). I'm not sure this is really necessary, but it's the only thing which occurs to me in this case.

I could not spot any significant difference on performance but I guess I did not set enough products/variants/overrides -> this will surely be noticeable in production.

I'd say this is ready to go.

**Pull request merged**